src/drivers/fluid_pipewire.c

```diff
@@ -0,0 +1,277 @@
+ * Audio driver for PipeWire.
+ *
+ */
+
+#include "fluid_synth.h"
+#include "fluid_adriver.h"
+#include "fluid_settings.h"
+
+#if PIPEWIRE_SUPPORT
+
+#include <pipewire/pipewire.h>
+#include <spa/param/audio/format-utils.h>
+
+/* At the moment, only stereo is supported */
+#define NUM_CHANNELS 2
+
+typedef struct
+{
+    fluid_audio_driver_t driver;
+    fluid_audio_func_t user_callback;
+    void *data;
+
+    /* Used only with the user-provided callback */
+    float *lbuf, *rbuf;
+
+    int buffer_period;
+    struct spa_pod_builder *builder;
+
+    struct pw_thread_loop *pw_loop;
+    struct pw_stream *pw_stream;
+} fluid_pipewire_audio_driver_t;
+
+/* Fast-path rendering routine with no user processing callbacks */
+static void fluid_pipewire_event_process(void *data)
+{
+    fluid_pipewire_audio_driver_t *drv = data;
+    struct pw_buffer *pwb;
+    struct spa_buffer *buf;
+    float *dest;
+    int stride;
+
+    pwb = pw_stream_dequeue_buffer(drv->pw_stream);
+    if (!pwb)
+    {
+        FLUID_LOG(FLUID_WARN, "No buffers!");
+        return;
+    }
+
+    buf = pwb->buffer;
+    dest = buf->datas[0].data;
+    if (!dest)
+    {
+        return;
+    }
+
+    stride = sizeof(float) * NUM_CHANNELS;
+
+    fluid_synth_write_float(drv->data, drv->buffer_period, dest, 0, 2, dest, 1, 2);
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->stride = stride;
+    buf->datas[0].chunk->size = drv->buffer_period * stride;
+
+    pw_stream_queue_buffer(drv->pw_stream, pwb);
+}
+
+/* Rendering routine with support for user-defined audio manipulation */
+static void fluid_pipewire_event_process2(void *data)
+{
+    fluid_pipewire_audio_driver_t *drv = data;
+    struct pw_buffer *pwb;
+    struct spa_buffer *buf;
+    float *dest;
+    float *channels[NUM_CHANNELS] = { drv->lbuf, drv->rbuf };
+    int stride, i;
+
+    pwb = pw_stream_dequeue_buffer(drv->pw_stream);
+    if (!pwb)
+    {
+        FLUID_LOG(FLUID_WARN, "No buffers!");
+        return;
+    }
+
+    buf = pwb->buffer;
+    dest = buf->datas[0].data;
+    if (!dest)
+    {
+        return;
+    }
+
+    stride = sizeof(float) * NUM_CHANNELS;
+
+    FLUID_MEMSET(drv->lbuf, 0, drv->buffer_period * sizeof(float));
+    FLUID_MEMSET(drv->rbuf, 0, drv->buffer_period * sizeof(float));
+
+    (*drv->user_callback)(drv->data, drv->buffer_period, 0, NULL, NUM_CHANNELS, channels);
+
+    /* Interleave the floating point data */
+    for (i = 0; i < drv->buffer_period; i++)
+    {
+        dest[i * 2] = drv->lbuf[i];
+        dest[i * 2 + 1] = drv->rbuf[i];
+    }
+
+    buf->datas[0].chunk->offset = 0;
+    buf->datas[0].chunk->stride = stride;
+    buf->datas[0].chunk->size = drv->buffer_period * stride;
+
+    pw_stream_queue_buffer(drv->pw_stream, pwb);
+}
+
+fluid_audio_driver_t *new_fluid_pipewire_audio_driver(fluid_settings_t *settings, fluid_synth_t *synth)
+{
+    return new_fluid_pipewire_audio_driver2(settings, NULL, synth);
+}
+
+static struct pw_stream_events events = {
```

Comment:
It is not clear to me why this needs to be static. Could you pls. add a comment?

Reply:
it doesn't, actually, let me fix this